### PR TITLE
Implement UDF Plan

### LINKED ISSUE #11242: Implement user defined planner for sql_position_to_expr
### Is your feature request related to a problem or challenge?

Part of #11207

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xinlifoobar**
## Which issue does this PR close?

Closes #11242 and part of #11207

### CODE REVIEW

File: datafusion/core/src/execution/session_state.rs [hunk 1]
```diff
@@ -967,14 +967,19 @@ impl SessionState {
...
+        #[cfg(feature = "unicode_expressions")]
         {
-            query
+            let position_planner =
+                Arc::new(functions::unicode::planner::PositionPlanner::default()) as _;
+
+            query = query.with_user_defined_planner(position_planner);
         }
+
+        query
```
Comment (dharanad) [inline datafusion/core/src/execution/session_state.rs:982]: As commented in https://github.com/apache/datafusion/pull/11243 let move this to constructor of `SessionState`

Comment (xinlifoobar) [reply datafusion/core/src/execution/session_state.rs:982]: Sorry this is not ready to review. See comment here.

https://github.com/apache/datafusion/issues/11207#issuecomment-2208578739

Comment (dharanad) [reply datafusion/core/src/execution/session_state.rs:982]: My bad, apologies. I comment as a precautionary measure ?

File: datafusion/sql/src/expr/mod.rs [hunk 2]
```diff
@@ -170,6 +170,7 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         schema: &DFSchema,
         planner_context: &mut PlannerContext,
     ) -> Result<Expr> {
+        let sql_not_moved = sql.clone();
```
Comment (xinlifoobar) [inline datafusion/sql/src/expr/mod.rs:173]: This isn't good. We may want something like https://github.com/VitWW/rfcs/blob/partial_types3/text/0000-partial_types.md

to implement partial move...

Comment (xinlifoobar) [reply datafusion/sql/src/expr/mod.rs:173]: Using references is also bad. The code moves fields from sqlparser::ast::Expr everywhere.

File: datafusion/sql/src/expr/mod.rs [hunk 3]
```diff
@@ -813,6 +795,37 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
     }
 
+    fn sql_udf_plan(
+        &self,
+        sql: SQLExpr,
+        schema: &DFSchema,
+        planner_context: &mut PlannerContext,
+    ) -> Result<Expr> {
+        let sql_not_moved = sql.clone();
+        let mut extracted_args = match sql {
```
Comment (jayzhan211) [inline datafusion/sql/src/expr/mod.rs:805]: it seems not to be extract specific anymore, we could name it `args`

**Review by jayzhan211 — COMMENTED** [review]
IMO I don't think this is a good abstration, because extract and position are actually quite different concept. For the point of view to user-defined planner, they could have their own implementation that may not be UDF, so it does not make sense to place them to plan_udf for me.

I would prefer to have `plan_extract` and `plan_position` separately.

Btw, I recommend `plan_udaf` in #11229, since we are customizing the query of aggregate function. Similarly, I think [here](https://github.com/apache/datafusion/blob/dce77db316beb5bf5dc2326c350b2d642edbfed8/datafusion/sql/src/expr/function.rs#L230-L233) should be the correct place for `plan_udf`

@xinlifoobar What do you think?

**Comment by alamb** [comment]
> I would prefer to have `plan_extract` and `plan_position` separately.

Here is a PR for planning position: https://github.com/apache/datafusion/pull/11243

It isn't clear to me what we should do with this PR now (should we close it?)

**Comment by alamb** [comment]
> IMO I don't think this is a good abstration, because extract and position are actually quite different concept. For the point of view to user-defined planner, they could have their own implementation that may not be UDF, so it does not make sense to place them to plan_udf for me.
> 
> I would prefer to have `plan_extract` and `plan_position` separately.

I agree that even though the functions are somewhat repetitively named, it is better to be explicit in the function names rather than being implicit (via function name) as the API proposed in this PR would be

Sorry I didn't see the comments on https://github.com/apache/datafusion/issues/11207#issuecomment-2210845458 on https://github.com/apache/datafusion/issues/11207 earlier

**Pull request closed without merging**